core/trino-main/src/main/java/io/trino/operator/FlatHash.java

```diff
@@ -0,0 +1,470 @@
+                int step = 1;
+                while (true) {
+                    final long controlVector = (long) LONG_HANDLE.get(control, bucket);
+                    // values are already distinct, so just find the first empty slot
+                    int emptyIndex = findEmptyInVector(controlVector, bucket);
+                    if (emptyIndex >= 0) {
+                        setControl(emptyIndex, hashPrefix);
+
+                        byte[] records = getRecords(emptyIndex);
+                        int recordOffset = getRecordOffset(emptyIndex);
+                        int oldRecordOffset = getRecordOffset(oldIndex);
+                        System.arraycopy(oldRecords, oldRecordOffset, records, recordOffset, recordSize);
+
+                        int groupId = (int) INT_HANDLE.get(records, recordOffset + recordGroupIdOffset);
+                        groupRecordIndex[groupId] = emptyIndex;
+
+                        break;
+                    }
+
+                    bucket = bucket(bucket + step);
+                    step += VECTOR_LENGTH;
+                }
+            }
+        }
+
+        // release temporary memory reservation
+        rehashMemoryReservation = 0;
+        checkMemoryReservation.update();
+    }
+
+    private int bucket(int hash)
+    {
+        return hash & mask;
+    }
+
+    private byte[] getRecords(int index)
+    {
+        return recordGroups[index >> RECORDS_PER_GROUP_SHIFT];
+    }
+
+    private int getRecordOffset(int index)
+    {
+        return (index & RECORDS_PER_GROUP_MASK) * recordSize;
+    }
+
+    private long valueHashCode(byte[] records, int index)
+    {
+        int recordOffset = getRecordOffset(index);
+
+        try {
+            byte[] variableWidthChunk = EMPTY_CHUNK;
+            if (variableWidthData != null) {
+                variableWidthChunk = variableWidthData.getChunk(records, recordOffset);
+            }
+
+            return flatHashStrategy.hash(records, recordOffset + recordValueOffset, variableWidthChunk);
+        }
+        catch (Throwable throwable) {
+            throwIfUnchecked(throwable);
+            throw new RuntimeException(throwable);
+        }
+    }
+
+    private boolean valueNotDistinctFrom(int leftIndex, Block[] rightBlocks, int rightPosition, long rightHash)
+    {
+        byte[] leftRecords = getRecords(leftIndex);
+        int leftRecordOffset = getRecordOffset(leftIndex);
+
+        if (hasPrecomputedHash) {
+            long leftHash = (long) LONG_HANDLE.get(leftRecords, leftRecordOffset + recordHashOffset);
+            if (leftHash != rightHash) {
+                return false;
+            }
+        }
+
+        byte[] leftVariableWidthChunk = EMPTY_CHUNK;
+        if (variableWidthData != null) {
+            leftVariableWidthChunk = variableWidthData.getChunk(leftRecords, leftRecordOffset);
+        }
+
+        return flatHashStrategy.valueNotDistinctFrom(
```

Comment:
every access to the `flatHashStrategy` is going to be a virtual call. Did you consider isolating the whole `FlatHash` per list of hash types?

Reply:
I am aware of this.  It doesn't seem to affect performance much as the distinct check is only called once per potential collision, but collisions are rare due to the swiss table design (an additional 1 in 128 chance of fp), and in most cases we need to keep the full hash for output wich is 64 bit.

But this is an area to explore for follow up performance improvements.  I stopped tuning after I saw the big win, so we can get it in.  It might be as simple as merging this with the `matchInVector` method as that is really the core of the loop.

Reply:
to see the impact you would have to pollute the jit profile in the benchmark. I don't think `BenchmarkGroupByHash` does it but yeah, comparing to the improvement here, the impact is probably not that big, also becausethe  current `MultiChannelGroupByHash` has the same interaction with `PagesHashStrategy`